File: packages/types/borrow.ts
import { Book } from './book'

export type BorrowStatus = 'active' | 'returned' | 'overdue'

export type BorrowTransaction = {
  id: string
  user_id: string
  book_id: string
  borrowed_at: string
  due_date: string
  returned_at?: string
  status: BorrowStatus
  fine_amount?: number
  /** Embedded via the books FK — present when the API selects it. */
  books?: Book
  /** Embedded via the user_id FK — the borrowing patron (id/email/full_name/role only). */
  profiles?: {
    id: string
    email: string
    full_name?: string
    role: string
  }
}

File: apps/api/schemas/borrow.py
from typing import Literal

from pydantic import BaseModel

from schemas.auth import UserProfile
from schemas.book import Book

BorrowStatus = Literal["active", "returned", "overdue"]

class BorrowTransaction(BaseModel):
    id: str
    user_id: str
    book_id: str
    borrowed_at: str
    due_date: str
    returned_at: str | None = None
    status: BorrowStatus
    fine_amount: float | None = None
    books: Book | None = None       # embedded via the books FK, when selected
    profiles: UserProfile | None = None  # embedded via the user_id FK — the borrowing patron

class CreateBorrowRequest(BaseModel):
    book_id: str
    patron_email: str
